src/modules/client/client.service.ts

```diff
@@ -0,0 +1,153 @@
+      .leftJoinAndSelect('freelancer.user', 'user')
+      .where({
+        category: {
+          id: +filters.category,
+        },
+      })
+      .getMany();
+
+    const filterKeyWords = filters.keyWords.toLowerCase().split(' ');
+
+    const filterSkills: number[] = filters.skills
+      .split(',')
+      .map((skill) => +skill);
+
+    const result = filterByCategory.filter((freelancer: Freelancer) => {
+      const resultKeyWordsFilter =
+        freelancer.user.description
+          .toLowerCase()
+          .split(' ')
+          .concat(freelancer.position.toLowerCase().split(' '))
+          .filter((descriptionWord: string) =>
+            filterKeyWords.includes(descriptionWord),
+          ).length !== 0;
+
+      const resultSkillsFilter =
+        freelancer.skills.filter((skill: Skill) =>
+          filterSkills.includes(skill.id),
+        ).length !== 0;
+
+      return resultKeyWordsFilter && resultSkillsFilter;
+    });
+
+    return await this.addSavesField(userId, result);
+  }
+
+  async getRecentlyViewedFreelancer(userId: number) {
+    const { recentlyViewedFreelancers } = await this.usersRepo
+      .createQueryBuilder('user')
+      .leftJoinAndSelect('user.recentlyViewedFreelancers', 'freelancer')
+      .leftJoinAndSelect('freelancer.user', 'freelancerUser')
+      .where({ id: userId })
+      .getOne();
+
+    return await this.addSavesField(userId, recentlyViewedFreelancers);
+  }
+
+  async viewFreelancer(userId: number, freelancerId: number) {
+    const freelancer = await this.freelancersRepo.findOneBy({
+      id: freelancerId,
+    });
+
+    const user = await this.usersRepo
+      .createQueryBuilder('user')
+      .leftJoinAndSelect('user.recentlyViewedFreelancers', 'freelancer')
+      .leftJoinAndSelect('freelancer.user', 'freelancerUser')
+      .where({ id: userId })
+      .getOne();
+
+    user.recentlyViewedFreelancers.push(freelancer);
+
+    this.usersRepo.save(user);
+    return await this.addSavesField(userId, user.recentlyViewedFreelancers);
+  }
+
+  async getSavedFreelancers(userId: number) {
+    const { savedFreelancers } = await this.usersRepo
+      .createQueryBuilder('user')
+      .leftJoinAndSelect('user.savedFreelancers', 'savedFreelancers')
+      .leftJoinAndSelect('savedFreelancers.user', 'savedFreelancersUser')
+      .where({ id: userId })
+      .getOne();
+
+    return savedFreelancers.map((freelancer) => {
+      return { saved: true, ...freelancer };
+    });
+  }
+
+  async saveFreelancer(userId: number, freelancerUserId: number) {
+    const freelancer = await this.freelancersRepo.findOneBy({
+      userId: freelancerUserId,
+    });
+
+    const user = await this.usersRepo
+      .createQueryBuilder('user')
+      .leftJoinAndSelect('user.savedFreelancers', 'savedFreelancers')
+      .where({ id: userId })
+      .getOne();
+    if (
+      user.savedFreelancers.filter(
+        (freelancer) => freelancer.userId == freelancerUserId,
```

Comment:
dont use ==

Reply:
you make filter twice, move it to the constant

Reply:
done